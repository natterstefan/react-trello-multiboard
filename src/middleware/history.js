import { push } from 'connected-react-router'
import { actions as appActions } from '../actions/app'
import { actions as boardActions } from '../actions/boards'

export default store => next => action => {
  switch (action.type) {
    case appActions.TOGGLE_PREFERRED_LIST:
      store.dispatch(push(`/pattern/${encodeURIComponent(action.pattern)}`))
      return next(action)

    case appActions.TOGGLE_PREFERRED:
    case boardActions.RESET_BOARDS:
      store.dispatch(push('/'))
      return next(action)

    default:
      return next(action)
  }
}
