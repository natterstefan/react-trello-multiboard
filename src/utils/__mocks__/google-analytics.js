export const initGA = jest.fn()
