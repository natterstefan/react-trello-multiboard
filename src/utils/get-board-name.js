import { findIndex, get } from 'lodash'
import Config from '../../config/config'

export const getBoardName = (name = '') => {
  const index = findIndex(Config.boards, b => b.name === name)
  return get(Config, `boards[${index}].shortcut`, name)
}
