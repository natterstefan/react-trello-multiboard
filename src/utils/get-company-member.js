import { find } from 'lodash'
import { company_member as companyMember } from '../../config/config'

const getCompanyMember = members => find(members, { username: companyMember })

export default getCompanyMember
