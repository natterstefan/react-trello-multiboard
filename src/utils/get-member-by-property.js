import { find } from 'lodash'

export const getMemberByOneOfProperty = (members, props = [], value) =>
  find(members, m => find(props, p => m[p] === value))
