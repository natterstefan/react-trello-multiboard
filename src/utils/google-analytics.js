/* eslint-disable */
import Config from '../../config/config'

/**
 * initGA
 *
 * initialises Google Analytics tracking (if set in Config), anonymizes the IP
 * triggers initial pageview and pageviews on consequent history changes
 */
export const initGA = history => {
  if (!Config.google_analytics_property) {
    // tracking is not enabled
    return
  }
  const gaProperty = Config.google_analytics_property
  const disableStr = 'ga-disable-' + gaProperty

  function gaOptout() {
    document.cookie = disableStr + '=true; expires=Thu, 31 Dec 2099 23:59:59 UTC; path=/'
    window[disableStr] = true
    alert('Google Analytics tracking was deactivated. Please refresh the page.')
  }
  window.gaOptout = gaOptout
  if (document.cookie.indexOf(disableStr + '=true') > -1) {
    window[disableStr] = true
    return // do not track
  }

  // always respect doNotTrack browser setting
  if (navigator.doNotTrack == '1' || window.doNotTrack == '1') {
    ;(function(i, s, o, g, r, a, m) {
      i['GoogleAnalyticsObject'] = r
      ;(i[r] =
        i[r] ||
        function() {
          ;(i[r].q = i[r].q || []).push(arguments)
        }),
        (i[r].l = 1 * new Date())
      ;(a = s.createElement(o)), (m = s.getElementsByTagName(o)[0])
      a.async = 1
      a.src = g
      m.parentNode.insertBefore(a, m)
    })(window, document, 'script', 'https://www.google-analytics.com/analytics.js', 'ga')
  }
  if (typeof ga === 'function') {
    ga('create', gaProperty, 'auto')
    ga('set', 'anonymizeIp', 1)
    ga('send', 'pageview', history.location.pathname) // initial pageview

    history.listen(location => {
      ga('send', 'pageview', location.pathname)
    })
  }
}
