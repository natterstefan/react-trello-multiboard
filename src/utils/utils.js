// cpontains sever random little helpers
import { compact, forOwn, get, map, some } from 'lodash'

export const checkState = (obj, prop, lookupProp) => {
  const results = []
  forOwn(get(obj, prop) || {}, value => {
    results.push(value && value[lookupProp])
  })
  return compact(results)
}

// get a list of isLoading states
export const isAppLoading = state => ({
  isBoardsLoading: get(state, 'boards.isLoading') || false,
  isCardsLoading: some(checkState(state, 'cards', 'isLoading'), i => i === true) || false,
  isListsLoading: some(checkState(state, 'lists', 'isLoading'), i => i === true) || false,
  isMembersLoading: get(state, 'members.isLoading') || false,
})

// get a list of errors states (list of errors or empty array when no error exists)
export const getAppErrorsList = state => {
  const getMessage = item => {
    if (!item) {
      return ''
    } else if (item.message) {
      return item.message
    }
    return map(item, i => i.message)
  }

  const errors = {
    boardError: get(state, 'boards.error') || null,
    cardError: checkState(state, 'cards', 'error') || null,
    listError: checkState(state, 'lists', 'error') || null,
    memberError: get(state, 'members.error') || null,
    userError: get(state, 'user.error') || null,
  }
  const appErrors = []
  forOwn(errors, (value, key) => {
    const message = getMessage(value)
    if (message.length === 0) {
      return null
    }

    return value && appErrors.push({ key, message })
  })

  return appErrors
}
