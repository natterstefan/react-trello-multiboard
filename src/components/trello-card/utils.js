import { get, includes, some } from 'lodash'
import { getMemberByOneOfProperty } from '../../utils/get-member-by-property'

/**
 * isCardHidden: finds out if the card is hidden or not (eg. when one member of
 * the card is toggled in the search)
 *
 * the current setup is only working in combination with the TrelloCardContainer
 * see /src/components/trello-card/index.js
 */
export const isCardHidden = (props = {}) => {
  const { memberToggle = {}, members = [], cardMembers = [] } = props
  let isShown = true // default: card is not hidden
  const togglePreferred = get(memberToggle, 'togglePreferred')
  const togglePreferredMember = get(memberToggle, 'togglePreferredMember')

  if (togglePreferred) {
    if (togglePreferredMember) {
      // first find the member and then check if the member is part of the cardMembers
      const foundMember =
        getMemberByOneOfProperty(members, ['id', 'username'], togglePreferredMember) || {}
      const foundMemberId = foundMember.id

      isShown = some(cardMembers, id => id === foundMemberId)
    } else {
      // members => all members we received and stored in the redux store
      // cardMembers => id's of the members on the current card
      isShown = some(members, member => includes(cardMembers, member.id))
    }
  }

  return !isShown
}
