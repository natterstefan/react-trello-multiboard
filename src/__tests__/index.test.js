import React from 'react'
import { shallow } from 'enzyme'
import { initGA } from '../utils/google-analytics'
import history from '../utils/history'

describe('Component/TrelloMultiBoard', () => {
  const ReactDom = require('react-dom')
  let TrelloMultiBoard

  beforeAll(() => {
    // Mocks
    ReactDom.render = jest.fn()

    // import once everything is mocked
    TrelloMultiBoard = require('../index').TrelloMultiboard
  })

  afterAll(() => {
    jest.unmock('react-dom')
    ReactDom.render.mockReset()
  })

  test('should render without throwing an error', () => {
    expect(shallow(<TrelloMultiBoard />)).toMatchSnapshot()
  })

  test('should call initGA', () => {
    expect(initGA).toHaveBeenCalledTimes(1)
    expect(initGA).toHaveBeenCalledWith(history)
  })
})
