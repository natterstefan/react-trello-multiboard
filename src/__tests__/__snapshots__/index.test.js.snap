// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component/TrelloMultiBoard should render without throwing an error 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <div>
    <WithStyles(CssBaseline) />
    <HashRouter>
      <Switch>
        <Route
          component={[Function]}
          exact={true}
          path="/config"
        />
        <Route
          component={[Function]}
          path="/"
        />
      </Switch>
    </HashRouter>
  </div>
</Provider>
`;
